src/main/java/org/killbill/billing/plugin/stripe/StripePaymentPluginApi.java

```diff
@@ -299,22 +299,25 @@ public void addPaymentMethod(final UUID kbAccountId, final UUID kbPaymentMethodI
 
         final Map<String, Object> additionalDataMap;
         final String stripeId;
+        final String existingCustomerId = getCustomerIdNoException(kbAccountId, context);
         if (paymentMethodIdInStripe != null) {
             if ("payment_method".equals(objectType)) {
                 try {
                     final PaymentMethod stripePaymentMethod = PaymentMethod.retrieve(paymentMethodIdInStripe, requestOptions);
                     additionalDataMap = StripePluginProperties.toAdditionalDataMap(stripePaymentMethod);
-                    ImmutableMap<String, Object> params = ImmutableMap.of("payment_method", paymentMethodIdInStripe);
-                    stripeId = createStripeCustomer(kbAccountId, paymentMethodIdInStripe, stripePaymentMethod.getId(), params, requestOptions, allProperties, context);
+                    ImmutableMap<String, Object> params = ImmutableMap.of("payment_method", stripePaymentMethod.getId());
+                    createStripeCustomer(kbAccountId, existingCustomerId, params, requestOptions, allProperties, context);
+                    stripeId = stripePaymentMethod.getId();
                 } catch (final StripeException e) {
                     throw new PaymentPluginApiException("Error calling Stripe while adding payment method", e);
                 }
             } else if ("token".equals(objectType)) {
                 try {
                     final Token stripeToken = Token.retrieve(paymentMethodIdInStripe, requestOptions);
                     additionalDataMap = StripePluginProperties.toAdditionalDataMap(stripeToken);
-                    ImmutableMap<String, Object> params = ImmutableMap.of("source", paymentMethodIdInStripe);
-                    stripeId = createStripeCustomer(kbAccountId, paymentMethodIdInStripe, stripeToken.getId(), params, requestOptions, allProperties, context);
+                    ImmutableMap<String, Object> params = ImmutableMap.of("source", stripeToken.getId());
+                    String customerId = createStripeCustomer(kbAccountId, existingCustomerId, params, requestOptions, allProperties, context);
+                    stripeId = retrievePaymentMethod(customerId, existingCustomerId, stripeToken.getId(), requestOptions);
```

Comment:
Here we want to add a Stripe Token payment method and we have multiple scenarios:

- The Kill Bill existingCustomerId is null  => we are going to create a Stripe Customer and the token will be [set as the default source](https://stripe.com/docs/api/customers/create#create_customer-source)
`// The id to charge now is the default source (e.g. card), not the token`
- Kill Bill has already an existingCustomerId => the Stripe Customer is not updated and the [Card ID is part of the Token response](https://stripe.com/docs/api/tokens/retrieve)

As per the comment `// The id to charge now is the default source (e.g. card), not the token` I understand that the token ID must not be used (`tok_xxxx`) because it is only for client side purposes. We must prefer the `card_xxx` or `ba_xxx` IDs

Reply:
I see - it works because `existingCustomerId == null`. Subtle!